cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/MacroParser.java

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+
+package co.cask.cdap.internal.app.runtime.plugin;
+
+
+import co.cask.cdap.api.macro.InvalidMacroException;
+import co.cask.cdap.api.macro.MacroEvaluator;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.regex.Pattern;
+import javax.annotation.Nullable;
+
+/**
+ * This class provides static utility functions for parsing macro syntax including:
+ *    - Maco Substitution
+ *    - Syntax validation
+ *    TODO: - Geting a list of macros a user must configure in a String
+ */
+public final class MacroParser {
+  private static final String[] ESCAPED_TOKENS = {"\\${", "\\}", "\\(", "\\)"};
+  private static final Pattern COMMA = Pattern.compile(",");
+  private static final int MAX_SUBSTITUTION_DEPTH = 10;
```

Comment:
If a macro has a circular substitution (like if "key" is replaced with ${key}) then expanding ${key} would lead to an infinite loop.